src/Microsoft.AspNet.WebUtilities/Encoders/UnicodeHelpers.cs

```diff
@@ -0,0 +1,164 @@
+
+        /// <summary>
+        /// Returns a bitmap of all characters which are defined per version 7.0.0
+        /// of the Unicode specification.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static uint[] GetDefinedCharacterBitmap()
+        {
+            return Volatile.Read(ref _definedCharacterBitmap) ?? CreateDefinedCharacterBitmap();
+        }
+
+        /// <summary>
+        /// Given a UTF-16 character stream, reads the next scalar value from the stream.
+        /// Set 'endOfString' to true if 'pChar' points to the last character in the stream.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int GetScalarValueFromUtf16(char* pChar, bool endOfString)
+        {
+            // This method is marked as AggressiveInlining to handle the common case of a non-surrogate
+            // character. The surrogate case is handled in the slower fallback code path.
+            char thisChar = *pChar;
+            return (Char.IsSurrogate(thisChar)) ? GetScalarValueFromUtf16Slow(pChar, endOfString) : thisChar;
+        }
+
+        private static int GetScalarValueFromUtf16Slow(char* pChar, bool endOfString)
+        {
+            char firstChar = pChar[0];
+
+            if (!Char.IsSurrogate(firstChar))
+            {
+                Debug.Fail("This case should've been handled by the fast path.");
+                return firstChar;
+            }
+            else if (Char.IsHighSurrogate(firstChar))
+            {
+                if (endOfString)
+                {
+                    // unmatched surrogate - substitute
+                    return UNICODE_REPLACEMENT_CHAR;
+                }
+                else
+                {
+                    char secondChar = pChar[1];
+                    if (Char.IsLowSurrogate(secondChar))
+                    {
+                        // valid surrogate pair - extract codepoint
+                        return GetScalarValueFromUtf16SurrogatePair(firstChar, secondChar);
+                    }
+                    else
+                    {
+                        // unmatched surrogate - substitute
+                        return UNICODE_REPLACEMENT_CHAR;
+                    }
+                }
+            }
+            else
+            {
+                // unmatched surrogate - substitute
+                Debug.Assert(Char.IsLowSurrogate(firstChar));
+                return UNICODE_REPLACEMENT_CHAR;
+            }
+        }
+
+        private static int GetScalarValueFromUtf16SurrogatePair(char highSurrogate, char lowSurrogate)
+        {
+            Debug.Assert(Char.IsHighSurrogate(highSurrogate));
+            Debug.Assert(Char.IsLowSurrogate(lowSurrogate));
+
+            // See http://www.unicode.org/versions/Unicode6.2.0/ch03.pdf, Table 3.5 for the
+            // details of this conversion. We don't use Char.ConvertToUtf32 because its exception
```

Comment:
Likely not worthwhile. Their API is perfectly fine for its intended use case; our use case is slightly different.